### PR TITLE
Fixes #13679 - fix content host migration.

### PR DESCRIPTION
**Pull request by waldenraines**
There have been a few changes since this migration was written
that have gone into master.  This commit fixes these issues so
that this migration can still run correctly.

http://projects.theforeman.org/issues/13679

### CODE REVIEW

**Comment by parthaa** [comment]
So this migration needs some important changes -> look at http://projects.theforeman.org/issues/13266

File: db/migrate/20150930183738_migrate_content_hosts.rb [hunk 1]
```diff
@@ -1,14 +1,19 @@
 class MigrateContentHosts < ActiveRecord::Migration
-  class Katello::HostCollectionAssociation < ActiveRecord::Base
+  class Katello::SystemRepository < ActiveRecord::Base
```
Comment (parthaa) [inline db/migrate/20150930183738_migrate_content_hosts.rb]: This is going to modify the actual Katello::SystemRepository class. It is better to just remove "Katello::" SystemRepository and set the table name similar to the template in http://projects.theforeman.org/issues/13266

File: db/migrate/20150930183738_migrate_content_hosts.rb [hunk 2]
```diff
@@ -1,14 +1,19 @@
...
+    belongs_to :system, :inverse_of => :system_repositories, :class_name => 'Katello::System'
+    belongs_to :repository, :inverse_of => :system_repositories, :class_name => 'Katello::Repository'
+  end
+
+  class Katello::SystemHostCollections < ActiveRecord::Base
     self.table_name = "katello_system_host_collections"
   end
 
-  class Katello::ErrataAssociation < ActiveRecord::Base
+  class Katello::SystemErratum < ActiveRecord::Base
```
Comment (parthaa) [inline db/migrate/20150930183738_migrate_content_hosts.rb]: Same here chuck the Katello::

File: db/migrate/20150930183738_migrate_content_hosts.rb [hunk 3]
```diff
@@ -1,14 +1,19 @@
...
     self.table_name = "katello_system_errata"
-  end
 
-  class Katello::RepositoryAssociation < ActiveRecord::Base
-    self.table_name = "katello_system_repositories"
+    belongs_to :system, :inverse_of => :system_errata, :class_name => 'Katello::System'
+    belongs_to :erratum, :inverse_of => :system_errata, :class_name => 'Katello::Erratum'
   end
 
   class Katello::System < ActiveRecord::Base
```
Comment (parthaa) [inline db/migrate/20150930183738_migrate_content_hosts.rb]: Same here chuck the katello

**Comment by parthaa** [comment]
I was recently addressing http://projects.theforeman.org/issues/13266 before pulp 2.8 changed my priorities . Here is what I did -> https://github.com/parthaa/katello/commit/6cb7df9f44a52cad290e10abda4d8a5e299b52be 
Migrate content host in that commit was only halfway there ... But I am just trying to point out the approach.

File: db/migrate/20150930183738_migrate_content_hosts.rb (hunk 1 above)
Comment (waldenraines) [inline db/migrate/20150930183738_migrate_content_hosts.rb]: If I do that what happens when I call code from my migration that in turn calls `Katello::Repository.some_method()`?

File: db/migrate/20150930183738_migrate_content_hosts.rb (hunk 3 above)
Comment (daviddavis) [inline db/migrate/20150930183738_migrate_content_hosts.rb]: This is problematic in that system has a type column so even if we try to redefine another class (e.g. MyMigrationSystem), calls to the database (e.g. MyMigrationSystem.first) are still going to return Katello::System objects. The only thing I can think of is call `becomes` on everything that comes out of the database (e.g. MyMigrationSystem.first.becomes(MyMigrationSystem)). Thoughts?

Comment (daviddavis) [inline db/migrate/20150930183738_migrate_content_hosts.rb]: Ping @ehelms @jlsherrill in case you have any thoughts.

Comment (daviddavis) [inline db/migrate/20150930183738_migrate_content_hosts.rb]: It looks like `Hypervisor`s are the only thing extending systems. Are we using `Hypervisor`?

**Comment by waldenraines** [comment]
@parthaa, @daviddavis, @jlsherrill updated.

This is going to require extensive testing as I basically rewrote all of the data parts of the migration.  I tried my best to test it out but I'm scared :fearful:

**Comment by stbenjam** [comment]
Get this error when migrating:


```
2016-02-12T15:00:45 [app] [I] Processing content host b7ff8c7a-709e-451a-8a8a-e9efc64e89ec katello-centos7-devel.example.com
2016-02-12T15:00:45 [app] [I] Found host with hostname katello-centos7-devel.example.com.
2016-02-12T15:00:45 [app] [I] Creating content facet for host katello-centos7-devel.example.com.
rake aborted!
StandardError: An error has occurred, this and all later migrations canceled:

Katello::Host::ContentFacet(#55773660) expected, got MigrateContentHosts::ContentFacet(#68725900)/home/vagrant/.rvm/gems/ruby-2.1.5/gems/activerecord-4.1.5/lib/active_record/associations/association.rb:216:in `r
aise_on_type_mismatch!'
/home/vagrant/.rvm/gems/ruby-2.1.5/gems/activerecord-4.1.5/lib/active_record/associations/has_one_association.rb:25:in `replace'
/home/vagrant/.rvm/gems/ruby-2.1.5/gems/activerecord-4.1.5/lib/active_record/associations/singular_association.rb:17:in `writer'
/home/vagrant/.rvm/gems/ruby-2.1.5/gems/activerecord-4.1.5/lib/active_record/associations/builder/association.rb:118:in `content_facet='
/home/vagrant/katello/db/migrate/20150930183738_migrate_content_hosts.rb:171:in `create_content_facet'
/home/vagrant/katello/db/migrate/20150930183738_migrate_content_hosts.rb:329:in `block in up'
/home/vagrant/.rvm/gems/ruby-2.1.5/gems/activerecord-4.1.5/lib/active_record/relation/delegation.rb:46:in `each'
/home/vagrant/.rvm/gems/ruby-2.1.5/gems/activerecord-4.1.5/lib/active_record/relation/delegation.rb:46:in `each'
/home/vagrant/katello/db/migrate/20150930183738_migrate_content_hosts.rb:295:in `up'
/home/vagrant/.rvm/gems/ruby-2.1.5/gems/activerecord-4.1.5/lib/active_record/migration.rb:598:in `exec_migration'
/home/vagrant/.rvm/gems/ruby-2.1.5/gems/activerecord-4.1.5/lib/active_record/migration.rb:579:in `block (2 levels) in migrate'
/home/vagrant/.rvm/gems/ruby-2.1.5/gems/activerecord-4.1.5/lib/active_record/migration.rb:578:in `block in migrate'
/home/vagrant/.rvm/gems/ruby-2.1.5/gems/activerecord-4.1.5/lib/active_record/connection_adapters/abstract/connection_pool.rb:294:in `with_connection'
/home/vagrant/.rvm/gems/ruby-2.1.5/gems/activerecord-4.1.5/lib/active_record/migration.rb:577:in `migrate'
/home/vagrant/.rvm/gems/ruby-2.1.5/gems/activerecord-4.1.5/lib/active_record/migration.rb:752:in `migrate'
/home/vagrant/.rvm/gems/ruby-2.1.5/gems/activerecord-4.1.5/lib/active_record/migration.rb:992:in `block in execute_migration_in_transaction'
/home/vagrant/.rvm/gems/ruby-2.1.5/gems/activerecord-4.1.5/lib/active_record/migration.rb:1038:in `block in ddl_transaction'
/home/vagrant/.rvm/gems/ruby-2.1.5/gems/activerecord-4.1.5/lib/active_record/connection_adapters/abstract/database_statements.rb:201:in `block in transaction'
/home/vagrant/.rvm/gems/ruby-2.1.5/gems/activerecord-4.1.5/lib/active_record/connection_adapters/abstract/database_statements.rb:209:in `within_new_transaction'
/home/vagrant/.rvm/gems/ruby-2.1.5/gems/activerecord-4.1.5/lib/active_record/connection_adapters/abstract/database_statements.rb:201:in `transaction'
/home/vagrant/.rvm/gems/ruby-2.1.5/gems/activerecord-4.1.5/lib/active_record/transactions.rb:208:in `transaction'
/home/vagrant/.rvm/gems/ruby-2.1.5/gems/activerecord-4.1.5/lib/active_record/migration.rb:1038:in `ddl_transaction'
/home/vagrant/.rvm/gems/ruby-2.1.5/gems/activerecord-4.1.5/lib/active_record/migration.rb:991:in `execute_migration_in_transaction'
/home/vagrant/.rvm/gems/ruby-2.1.5/gems/activerecord-4.1.5/lib/active_record/migration.rb:953:in `block in migrate'
/home/vagrant/.rvm/gems/ruby-2.1.5/gems/activerecord-4.1.5/lib/active_record/migration.rb:949:in `each'
/home/vagrant/.rvm/gems/ruby-2.1.5/gems/activerecord-4.1.5/lib/active_record/migration.rb:949:in `migrate'
/home/vagrant/.rvm/gems/ruby-2.1.5/gems/activerecord-4.1.5/lib/active_record/migration.rb:807:in `up'
/home/vagrant/.rvm/gems/ruby-2.1.5/gems/activerecord-4.1.5/lib/active_record/migration.rb:785:in `migrate'
/home/vagrant/.rvm/gems/ruby-2.1.5/gems/activerecord-4.1.5/lib/active_record/railties/databases.rake:34:in `block (2 levels) in <top (required)>'
/home/vagrant/.rvm/gems/ruby-2.1.5/bin/ruby_executable_hooks:15:in `eval'
/home/vagrant/.rvm/gems/ruby-2.1.5/bin/ruby_executable_hooks:15:in `<main>'
ActiveRecord::AssociationTypeMismatch: Katello::Host::ContentFacet(#55773660) expected, got MigrateContentHosts::ContentFacet(#68725900)
```

**Comment by waldenraines** [comment]
> Get this error when migrating:

Odd, this works for me:

```
2016-02-12T15:06:21 [app] [I] Processing content host 538f6e75-e3cb-4fb8-bae7-a9cb4c442cac walden-rhel7.rdu.redhat.com
2016-02-12T15:06:21 [app] [I] Found host with hostname walden-rhel7.rdu.redhat.com.
2016-02-12T15:06:21 [app] [I] Creating content facet for host walden-rhel7.rdu.redhat.com.
2016-02-12T15:06:21 [app] [I] Creating subscription facet for host walden-rhel7.rdu.redhat.com.
```

**Comment by waldenraines** [comment]
I was able to reproduce @stbenjam's issue by not defining `Host` in the migration (since his managed host class overrides my defined host class for the host in question).

We are left with these options:

1. Do not redefine the host class, the content facet class, and the subscription facet classes.
1. Redefine the host class and go down the rabbit hole of attributes and methods it requires
1. Some other ruby magic that I don't know

@jlsherrill @daviddavis @parthaa @stbenjam thoughts?

**Comment by waldenraines** [comment]
@stbenjam can you try the latest please?

**Comment by stbenjam** [comment]
Same problem, I don't think overriding the type column fools rails, it still bails on Foreman's overridden becomes method:

```
NoMethodError: undefined method `drop_primary_interface_cache' for #<MigrateContentHosts::Host:0x0000001280a568>
```

**Comment by waldenraines** [comment]
> I was able to reproduce @stbenjam's issue by not defining Host in the migration (since his managed host class overrides my defined host class for the host in question).
> 
> We are left with these options:
> 
> Do not redefine the host class, the content facet class, and the subscription facet classes.
> Redefine the host class and go down the rabbit hole of attributes and methods it requires (I am trying this method currently but it's still failing because foreman overrides the becomes method )
> Some other ruby magic that I don't know
> @jlsherrill @daviddavis @parthaa @stbenjam thoughts?

Disregard this comment.  @parthaa was able to find that setting `self.inheritance_column = nil` causes the migration to not use STI.

**Comment by stbenjam** [comment]
All of my hosts/systems migrated with the current incarnation :+1:

**Comment by waldenraines** [comment]
[test]

**Comment by jlsherrill** [comment]
Current failure due to http://projects.theforeman.org/issues/13723

**Comment by jlsherrill** [comment]
ACK

**Comment by stbenjam** [comment]
:tada:

**Pull request merged**